python/private/pypi/dependency_resolver/dependency_resolver.py

```diff
@@ -185,14 +186,23 @@ def main(
             # and we should copy the updated requirements back to the source tree.
             if not absolute_output_file.samefile(requirements_file_tree):
                 atexit.register(
-                    lambda: shutil.copy(
-                        absolute_output_file, requirements_file_tree
-                    )
+                    lambda: shutil.copy(absolute_output_file, requirements_file_tree)
                 )
-        cli(argv, standalone_mode = False)
+        cli(argv, standalone_mode=False)
         requirements_file_relative_path = Path(requirements_file_relative)
         content = requirements_file_relative_path.read_text()
         content = content.replace(absolute_path_prefix, "")
+        # Replace the lines like: # via -r /absolute/path/to/requirements_extra.in
+        # with: # via -r requirements_extra.in
+        pattern = re.compile(r"(# via -r )(/.+/requirements.*\.in)")
+
+        def replace_abspath_with_name(match):
+            assert match is not None
+            # match[1] is "# via -r "
+            # match[2] is "/absolute/path/to/requirements*.in"
+            return match[1] + Path(match[2]).name
+
+        content = pattern.sub(replace_abspath_with_name, content)
```

Comment:
@aignas is this something like what you thought about in this message: https://github.com/bazelbuild/rules_python/pull/2514#discussion_r1895346149 ?

How about `UPDATE == False` case?

Reply:
I was actually thinking that the `absolute_path_prefix` replacement stopped working due to some reason and that we should fix the `absolute_path_prefix` matching, but it seems that in reality it is something else.

Would it be possible to have the `absolute_path_prefix` contain the actual absolute path prefix that works here?

Reply:
I'm not totally sure what does `absolute_path_prefix` replacement, running the test
```
cd tests/integration/compile_pip_requirements

echo '' > requirements_lock.txt && \
    bazel run //:requirements.update && \
    git diff --exit-code
```
**on `main`**, I have the following values in the vars:
```python
resolved_requirements_file: /tmp/jax/rules_python/tests/integration/compile_pip_requirements/requirements_lock.txt
absolute_path_prefix: /tmp/jax/rules_python/tests/integration/compile_pip_requirements/ 
```
However, requirements lock content does not contain these paths (even before `content = content.replace(absolute_path_prefix, "")` line):
```
# This file is autogenerated by pip-compile with Python 3.9                                                
# by the following command:                  
#                                                                                                          
#    bazel run //:requirements.update                                                                      
#                                                                                                          
                                                                                                           
# The following packages are considered to be unsafe in a requirements file:                               
pip==22.3.1 \
    --hash=sha256:65fd48317359f3af8e593943e6ae1506b66325085ea64b706a998c6e83eeaf38 \
    --hash=sha256:908c78e6bc29b676ede1c4d57981d490cb892eb45cd8c214ab6298125119e077
    # via -r requirements.in
setuptools==65.6.3 \
    --hash=sha256:57f6f22bde4e042978bcd50176fdb381d7c21a9efa4041202288d3737a0c6a54 \
    --hash=sha256:a7620757bf984b58deaf32fc8a4577a9bbc0850cf92c20e1ce41c38c19e5fb75
    # via -r requirements_extra.in
```
So, nothing is replaced with `content = content.replace(absolute_path_prefix, "")`.

If you can share more details on what was the intention of `content = content.replace(absolute_path_prefix, "")`, I can try to figure out whether this could work with never pip.

Reply:
The thing is that on Windows things work a little bit differently and it could be that this is only needed there. I can't fully remember the exact reasons for the code, but I do think that trying to use that might be useful.

Reply:
Let me recap the problem.

New pip==24.3.1 (vs 24.0) now adds the annotation as an absolute path vs relative path previously:
```diff
- # via -r requirements.in
+ # via -r /root/.cache/bazel/_bazel_root/a76a5e2851b1f2cc4ad9886a6a6aacc6/execroot/_main/bazel-out/k8-fastbuild/bin/requirements.update.runfiles/_main/requirements.in
```

In the test, for example, `tests/integration/compile_pip_requirements`, we use piptools compile as following:
```python
argv = ['--output-file=requirements_lock.txt', 'requirements.txt', '--resolver=backtracking', '--allow-unsafe', '--generate-hashes']
cli(argv)
```
Files `requirements.txt` and `requirements_lock.txt` are and will be located in `/root/.cache/bazel/_bazel_root/a76a5e2851b1f2cc4ad9886a6a6aacc6/execroot/_main/bazel-out/k8-fastbuild/bin/requirements.update.runfiles/_main/` folder.

The content of `requirements.txt` is:
```txt
-r requirements.in
```
After the execution, pip produces the lock file with the following content:
```txt
#
# This file is autogenerated by pip-compile with Python 3.9
# by the following command:
#
#    bazel run //:requirements.update
#

# The following packages are considered to be unsafe in a requirements file:
pip==22.3.1 \
    --hash=sha256:65fd48317359f3af8e593943e6ae1506b66325085ea64b706a998c6e83eeaf38 \
    --hash=sha256:908c78e6bc29b676ede1c4d57981d490cb892eb45cd8c214ab6298125119e077
    # via -r /root/.cache/bazel/_bazel_root/a76a5e2851b1f2cc4ad9886a6a6aacc6/execroot/_main/bazel-out/k8-fastbuild/bin/requirements.update.runfiles/_main/requirements.in
setuptools==65.6.3 \
    --hash=sha256:57f6f22bde4e042978bcd50176fdb381d7c21a9efa4041202288d3737a0c6a54 \
    --hash=sha256:a7620757bf984b58deaf32fc8a4577a9bbc0850cf92c20e1ce41c38c19e5fb75
    # via -r /root/.cache/bazel/_bazel_root/a76a5e2851b1f2cc4ad9886a6a6aacc6/execroot/_main/bazel-out/k8-fastbuild/bin/requirements.update.runfiles/_main/requirements_extra.in
```
And the test expects to see 
```
...
# via -r requirements.in
...
# via -r requirements_extra.in
```

On my linux machine, `absolute_path_prefix` is set to `/tmp/jax/rules_python/tests/integration/compile_pip_requirements/` where the test source code is located. It is computed from `resolved_requirements_file` which is again points to the original requirements file: `/tmp/jax/rules_python/tests/integration/compile_pip_requirements/requirements_lock.txt`

I'm not sure if we can efficiently use `absolute_path_prefix` without breaking something else (like test on Windows).

We can probably fetch the absolute path of the source `requirement.txt` file <-> `argv[1]` and remove it from all lines `# via -r ...`.  What do you think ?

Concerning my code in this PR, it may not always work as source file can be anything and not only `requirements*.in` as I'm regexping now...

Reply:
Thanks for the recap.

I guess we should ensure that neither source absolute paths nor the output absolute paths are in the requirements file. I think a solution that does not rely on `regexp` matching the filename could be good enough.